feat: Migrate to urfave/cli

sspaink opened the pull request:
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

related to: #11316

There is a goal to clean up the flags and commands of Telegraf (#11316), to help that effort I've migrated the project from using the [standard library flag package](https://pkg.go.dev/flag) to using [urfave/cli](https://github.com/urfave/cli). The urfave/cli package offers a clean and straightforward way to implement flags and commands. I believe this will help provide and easy foundation to begin updating and maintaining the flags/commands. 

The goal of this pull request is to just begin using "urfave/cli" package, with **no breaking changes**, so everything should work as it does now. Then in future pull requests, we can begin deprecating and introducing new flags/commands. I've also added test coverage around the flag/commands which didn't exist before. To add the test coverage I moved the main business logic to a separate function called `runApp` where the actual agent, pprof, and config get passed in. These are mocked out in the new tests.

The flags `--sample-config` and `--version` have already been marked as deprecated because there are duplicate commands that already exists  `telegraf config` and `telegraf version`. I also marked `--plugin-directory` as deprecated as this is a feature that was quickly disabled after introduction and would require a special build of Telegraf to get working that is not being released officially.

powersj commented:
Huge thank you for this. It will make things much more consistent and more easily make changes down the road. Some initial questions for our chat today after quickly scanning this PR:

1) Windows only flags now show up in linux output, is there a way to filter these? I don't find this as a blocker
2) Tested on mac and windows? What can I do to help?
3) Why were the internal version variables made public? 
4) What prompted the pprof changes?

sspaink commented:
**1. Windows only flags now show up in linux output, is there a way to filter these? I don't find this as a blocker**
Good catch, this is fixable
**2. Tested on mac and windows? What can I do to help?**
Windows services and Mac needs to be tested
**3. Why were the internal version variables made public?**
For testing, https://github.com/influxdata/telegraf/blob/cliv2/cmd/telegraf/telegraf_test.go#L314-L357 using it to set the version to something predictable.
**4.  What prompted the pprof changes?**
For testing and err channel was added to avoid calling panic if the pprof fails 


Action items from discussion:
1. Test windows
2. Test mac
3. Test error conditions (bad config ,etc)
4. Test version
5. Test sample config with diff
6. Add tests for all flags
7. Consider different name for "AgentManager",

powersj commented:
> Test sample config with diff

This looks good no diff between the file generated and the file from master

> Test version

Once I added your Makefile changes locally, I think the only difference is the lack of a newline.

```shell
# telegraf from your branch
❯ ./telegraf version | od -c
0000000   T   e   l   e   g   r   a   f       1   .   2   4   .   0   -
0000020   a   a   b   9   5   4   a   8       (   g   i   t   :       c
0000040   l   i   v   2   @   a   a   b   9   5   4   a   8   )
0000056
# this is telegraf built from master
❯ /tmp/telegraf version | od -c
0000000   T   e   l   e   g   r   a   f       1   .   2   4   .   0   -
0000020   9   e   a   2   d   e   a   c       (   g   i   t   :       m
0000040   a   s   t   e   r   @   9   e   a   2   d   e   a   c   )  \n
0000060

sspaink commented:
@powersj I've added the missing newline, nice find!

sspaink commented:
I've tested windows service by running the following commands:
```
telegraf.exe --service install	Install telegraf as a service
telegraf.exe --service uninstall	Remove the telegraf service
telegraf.exe --service start	Start the telegraf service
telegraf.exe --service stop	Stop the telegraf service
```

With a simple config:
```
[[inputs.cpu]]

[[outputs.file]]
    files = ["H:\\Sandbox\\test.txt"]
```

There was an issue I resolved in the latest commit, were I had added an error channel that was blocking the service from starting. I replaced it with just printing an error message.

The struct `AgentManager` has also been renamed to `Telegraf` to prevent overloading the word `Agent` which already has a meaning in this project.

powersj commented on cmd/telegraf/main.go:
```diff
@@ -0,0 +1,356 @@
+		},
+		&cli.StringFlag{
+			Name:  "processor-filter",
+			Usage: "filter the processors to enable, separator is :",
```
```suggestion
			Usage: "filter the processors to enable, separator is ':'",
```

powersj commented on cmd/telegraf/main.go:
```diff
@@ -0,0 +1,356 @@
+		},
+		&cli.StringFlag{
+			Name:  "aggregator-filter",
+			Usage: "filter the aggregators to enable, separator is :",
```
```suggestion
			Usage: "filter the aggregators to enable, separator is ':'",
```

powersj commented on cmd/telegraf/main.go:
```diff
@@ -0,0 +1,356 @@
+		},
+		&cli.StringFlag{
+			Name:  "output-filter",
+			Usage: "filter the outputs to enable, separator is :",
```
```suggestion
			Usage: "filter the outputs to enable, separator is ':'",
```

powersj commented on cmd/telegraf/main.go:
```diff
@@ -0,0 +1,356 @@
+		},
+		&cli.StringFlag{
+			Name:  "input-filter",
+			Usage: "filter the inputs to enable, separator is :",
```
```suggestion
			Usage: "filter the inputs to enable, separator is ':'",
```

powersj commented on cmd/telegraf/telegraf_posix.go:
```diff
@@ -3,10 +3,13 @@
-func run(inputFilters, outputFilters []string) {
+import "github.com/urfave/cli/v2"
+
+func (a *Telegraf) Run() error {
```
nit: but we renamed this from agent, but it is still referenced as `a` throughout

powersj reviewed (commented):
I completed my testing on macOS, went through each of the options, and ensured they worked as I understood them.

On Linux, I tried several invalid configs and options to ensure we still error as expected.

I have put a few comments below, primarily nits and clean-up.

powersj commented on cmd/telegraf/main.go:
```diff
@@ -0,0 +1,356 @@
+				},
+				&cli.StringFlag{
+					Name:  "pprof-addr",
+					Usage: "pprof address to listen on, not activate pprof if empty",
```
I forgot to click save on this comment...

`, not activate pprof if empty` - this doesn't sound clear enough. Essentially the user has to provide a `host|IP:port`to use for this option, right?

Maybe `pprof host/IP and port to listen on (e.g. 'localhost:8080')`?

sspaink replied:
Seems like a good change to me, I used localhost:6060 as the example because it matches the docs

powersj reviewed (approved):
+1

Looks great, and makes it much easier for us to maintain the CLI going forward, as well as extend it with additional subcommand and flags.

My only concern was re: internal variables, but we can chat as a team about that and make changes before v1.24.0 if needed.

reimda commented on cmd/telegraf/telegraf.go:
```diff
@@ -214,40 +159,37 @@ func watchLocalConfig(signals chan os.Signal, fConfig string) {
-	inputFilters []string,
-	outputFilters []string,
-) error {
+func (a *Telegraf) gather(ctx context.Context) error {
```
Why rename runAgent to gather? It doesn't look like it does anything different and it does more than just gather, like loading config, setting up logging.

sspaink replied:
Good catch, so orginially the struct was called `AgentManager` so I renamed it so it wasn't `agent.runAgent`, but that isn't a problem anymore. I renamed it back to `runAgent` like it was originally.

reimda commented on cmd/telegraf/telegraf.go:
```diff
@@ -3,138 +3,78 @@ package main
+}
 
-var stop chan struct{}
+type Telegraf struct {
```
It makes a lot of sense to have a struct that owns the vars it uses instead of them all being globals!

reimda commented on cmd/telegraf/main.go:
```diff
@@ -0,0 +1,352 @@
+				},
+				// !!!
+			}, extraFlags...),
+		Action: func(cCtx *cli.Context) error {
```
I like how urfave/cli uses structs nested in cli.App to configure cli options (flags and commands) but I think including funcs in the middle of the configuration structs (the `Action` fields) makes it harder to read when the func is long like it is here. The func's indentation is pretty deep here, five levels if I'm counting right.

What do you think about moving this func to be defined before the cli.App around line 85, then refer to it here by name? I think it will be one or two less indentation levels which should be a bit easier to read.

sspaink replied:
Seems like a good idea to me, less indentation is definitely better. I've moved the function to line 85
